### PR TITLE
[Algo] Attribute-based Representations for Accurate and Interpretable Video Anomaly Detection

### PR DESCRIPTION
**Pull request by djdameln**
# Description

- Algo implementation of [Attribute-based Representations for Accurate and Interpretable Video Anomaly Detection](https://paperswithcode.com/paper/attribute-based-representations-for-accurate).

Example predictions:

Known Issues:
- GMM module uses numpy and sklearn for now. This needs to be replaced by a custom PyTorch implementation in the future.
- The original implementation applies several post-processing operations to the bounding box predictions of the region extractor, which are not described in the paper. These operations are currently missing in my implementation.
- Gaussian temporal smoothing of the predictions is missing. This will need to be addressed separately, because it will involve some design (we could design this in a generic way so that it could be re-used for all spatio-temporal video anomaly models.)

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

### CODE REVIEW

File: src/anomalib/models/ai_vad/density.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    def predict(self, features: dict[FeatureType, Tensor]) -> tuple[Tensor, Tensor]:
+        """Predict the region- and image-level anomaly scores for an image based on a set of features.
+
+        Args:
+            features (dict[Tensor]): Dictionary containing extracted features for a single frame.
+        Returns:
+            Tensor: Region-level anomaly scores for all regions withing the frame.
+            Tensor: Frame-level anomaly score for the frame.
+        """
+        region_scores = torch.zeros(list(features.values())[0].shape[0]).to(list(features.values())[0].device)
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/density.py:117]: This line is a bit hard to follow. It might be an idea to add a comment here, or simplify.

File: src/anomalib/models/ai_vad/density.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            distances, _ = distances.min(1)
+            return distances.unsqueeze(1)
+        else:
+            distances, _ = distances.topk(k=n_neighbors, largest=False, dim=1)
+        return distances
+
+    def _compute_normalization_statistics(self):
+        """Compute min-max normalization statistics while taking the group into account."""
+        for group, features in self.memory_bank.items():
+            nns = self.predict(features, group, normalize=False)
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/density.py:221]: what does `nns` refer to?

File: src/anomalib/models/ai_vad/density.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        super().__init__()
+
+        self.gmm = GaussianMixture(n_components=n_components, random_state=0)
+        self.memory_bank: list[Tensor] | Tensor = []
+
+        self.normalization_statistics = MinMax()
+
+    def update(self, features: Tensor, _group: Any = None):
+        """Update the feature bank."""
+        self.memory_bank.append(features)
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/density.py:256]: below is the new recommended way to ignore unused variables

```suggestion
    def update(self, features: Tensor, group: Any = None):
        """Update the feature bank."""
        del group    # This variable is not needed.
        self.memory_bank.append(features)
```

File: src/anomalib/models/ai_vad/density.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    """Density estimation based on Gaussian Mixture Model.
+
+    Args:
+        n_components (int): Number of components used in the GMM.
+    """
+
+    def __init__(self, n_components: int = 2) -> None:
+        super().__init__()
+
+        self.gmm = GaussianMixture(n_components=n_components, random_state=0)
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/density.py:249]: Can you add a TODO and Jira reference here so that it would be on our agenda?

File: src/anomalib/models/ai_vad/features.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
+"""Feature extraction module for AI-VAD model implementation"""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from enum import Enum
+
+import clip
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:10]: is `clip` installed? I thought we are going to refer to a local file ?

File: src/anomalib/models/ai_vad/features.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        # Extract features
+        feature_dict = {}
+        if self.use_velocity_features:
+            velocity_features = self.velocity_extractor(flow_batch, boxes)
+            feature_dict[FeatureType.VELOCITY] = [velocity_features[indices == i] for i in range(batch_size)]
+        if self.use_pose_features:
+            pose_features = self.pose_extractor(rgb_batch, boxes_list)
+            feature_dict[FeatureType.POSE] = pose_features
+        if self.use_deep_features:
+            deep_features = self.appearance_extractor(rgb_batch, boxes, batch_size)
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:92]: Is `deep_features` the way how it is named in the paper?  Or `appearance_features`?

Comment (samet-akcay) [reply src/anomalib/models/ai_vad/features.py:92]: I'm a bit confused here. `self.appearance_extractor` is a `DeepExtractor` object. I feel like the naming would be more consistent?

File: src/anomalib/models/ai_vad/features.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+    def forward(self, batch: Tensor, boxes: Tensor, batch_size: int) -> Tensor:
+        """Extract deep features using CLIP encoder.
+
+        Args:
+            batch (Tensor): Batch of RGB input images of shape (N, 3, H, W)
+            boxes (Tensor): Bounding box coordinates of shaspe (M, 5). First column indicates batch index of the bbox.
+            batch_size (int): Number of images in the batch.
+        Returns:
+            Tensor: Deep feature tensor of shape (M, 512)
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:121]:
```suggestion
            batch_size (int): Number of images in the batch.

        Returns:
            Tensor: Deep feature tensor of shape (M, 512)
```

File: src/anomalib/models/ai_vad/features.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+        self.n_bins = n_bins
+
+    def forward(self, flows: Tensor, boxes: Tensor) -> Tensor:
+        """Extract velocioty features by filling a histogram.
+
+        Args:
+            flows (Tensor): Batch of optical flow images of shape (N, 2, H, W)
+            boxes (Tensor): Bounding box coordinates of shaspe (M, 5). First column indicates batch index of the bbox.
+        Returns:
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:153]:
```suggestion
        
        Returns:
```

File: src/anomalib/models/ai_vad/features.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+            Tensor: Velocity feature tensor of shape (M, n_bins)
+        """
+        flow_regions = roi_align(flows, boxes, output_size=[224, 224])
+
+        # cartesian to polar
+        mag_batch = torch.linalg.norm(flow_regions, axis=1, ord=2)
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:159]: what does `mag` refer to?

File: src/anomalib/models/ai_vad/features.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        theta_batch = torch.atan2(flow_regions[:, 0, ...], flow_regions[:, 1, ...])
+
+        # compute velocity histogram
+        velocity_hictograms = []
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:163]: Typo?
```suggestion
        velocity_histograms = []
```

File: src/anomalib/models/ai_vad/features.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+    @staticmethod
+    def _post_process(keypoint_detections: list[dict]) -> list[Tensor]:
+        """Convert keypoint predictions to 1D feature vectors.
+
+        Post-processing consists of flattening and normalizing to bbox coordinates.
+
+        Args:
+            keypoint_detections (list[dict]): Outputs of the keypoint extractor
+        Returns:
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:201]:
```suggestion

        Returns:
```

File: src/anomalib/models/ai_vad/features.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+            poses.append(normalized_keypoints.reshape(normalized_keypoints.shape[0], -1))
+        return poses
+
+    def forward(self, batch: Tensor, boxes: Tensor) -> list[Tensor]:
+        """Extract pose features using a human keypoint estimation model.
+
+        Args:
+            batch (Tensor): Batch of RGB input images of shape (N, 3, H, W)
+            boxes (Tensor): Bounding box coordinates of shaspe (M, 5). First column indicates batch index of the bbox.
+        Returns:
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/features.py:218]:
```suggestion

        Returns:
```

**Review by samet-akcay — COMMENTED** [review]
Great contribution! I've got few comments

File: src/anomalib/models/ai_vad/config.yaml [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  name: ucsd #options: [mvtec, btech, folder]
+  format: ucsdped
+  path: ./datasets/ucsd
+  category: UCSDped2
+  task: detection
+  clip_length_in_frames: 2
+  frames_between_clips: 10
+  video_target_frame: "last"
+  image_size: [240, 360] # RKDE region- and feature extractor apply custom resizing
+  normalization: none # RKDE region- and feature extractor apply custom normalization
```
[the remark this answers is not in the document]
Comment (djdameln) [reply src/anomalib/models/ai_vad/config.yaml:11]: Copy-pasted this part of the config. I'll update it

File: src/anomalib/models/ai_vad/config.yaml [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  use_deep_features: True
+  n_components_velocity: 2
+  n_neighbors_pose: 1
+  n_neighbors_deep: 1
+  # generic params
+  normalization_method: min_max # options: [null, min_max, cdf]
+
+metrics:
+  image:
+    - F1Score
```
[the remark this answers is not in the document]
Comment (djdameln) [reply src/anomalib/models/ai_vad/config.yaml:45]: In the paper they report Micro and Macro AUROC. We currently only support Micro. But I'll remove F1 for now.

File: src/anomalib/models/ai_vad/features.py (hunk 6 above)
Comment (djdameln) [reply src/anomalib/models/ai_vad/features.py:93]: I changed it from `appearance` to `deep`, to make it consistent with the paper, but missed this attribute apparently

File: src/anomalib/models/ai_vad/features.py (hunk 9 above)
Comment (djdameln) [reply src/anomalib/models/ai_vad/features.py:162]: magnitudes

File: src/anomalib/models/ai_vad/clip/LICENSE [hunk 15]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (ashwinvaidya17) [inline src/anomalib/models/ai_vad/clip/LICENSE:1]: Should we update `third-party-programs.txt` as well?

File: src/anomalib/data/ucsd_ped.py [hunk 16]
```diff
@@ -166,8 +167,9 @@ def __init__(
         split: Split,
         clip_length_in_frames: int = 1,
         frames_between_clips: int = 1,
+        target_frame: VideoTargetFrame = VideoTargetFrame.LAST,
```
Comment (ashwinvaidya17) [inline src/anomalib/data/ucsd_ped.py:170]: We should include this in the doc string as well

File: src/anomalib/models/ai_vad/__init__.py [hunk 17]
```diff
@@ -0,0 +1,13 @@
+"""Implementatation of the AI-VAD Model.
+
+AI-VAD: Accurate and Interpretable Video Anomaly Detection
+
+Paper https://arxiv.org/pdf/2212.00789.pdf
+"""
+
+# Copyright (C) 2022 Intel Corporation
```
Comment (ashwinvaidya17) [inline src/anomalib/models/ai_vad/__init__.py:8]: `2023`?

**Review by ashwinvaidya17 — COMMENTED** [review]
Very clean code! I have no major comments just few related to docstrings

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks for the efforts! Looks good to me

File: src/anomalib/models/ai_vad/clip/model.py [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+        self.class_embedding = nn.Parameter(scale * torch.randn(width))
+        self.positional_embedding = nn.Parameter(scale * torch.randn((input_resolution // patch_size) ** 2 + 1, width))
+        self.ln_pre = LayerNorm(width)
+
+        self.transformer = Transformer(width, layers, heads)
+
+        self.ln_post = LayerNorm(width)
+        self.proj = nn.Parameter(scale * torch.randn(width, output_dim))
+
+    def forward(self, x: torch.Tensor):
```
Comment (samet-akcay) [inline src/anomalib/models/ai_vad/clip/model.py:244]:
```suggestion
    def forward(self, x: torch.Tensor) -> torch.Tensor:
```

File: src/anomalib/models/ai_vad/clip/clip.py [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+#
+# Modified
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import hashlib
+import os
+import urllib
+import warnings
+from typing import List, Union
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply src/anomalib/models/ai_vad/clip/clip.py:16]: I feel we should treat everything under `clip` as third-party.

File: src/anomalib/models/padim/config.yaml [hunk 20]
```diff
@@ -59,7 +59,7 @@ project:
   path: ./results
 
 logging:
-  logger: [] # options: [comet, tensorboard, wandb, csv] or combinations.
+  logger: [wandb] # options: [comet, tensorboard, wandb, csv] or combinations.
```
Comment (samet-akcay) [inline src/anomalib/models/padim/config.yaml:62]: intentional?

Comment (djdameln) [reply src/anomalib/models/padim/config.yaml:62]: nope :)

**Review by samet-akcay — APPROVED** [review]
